packages/aws-cdk/lib/deploy.ts

```diff
@@ -0,0 +1,64 @@
+import * as cxapi from '@aws-cdk/cx-api';
+import PQueue from 'p-queue';
+
+type Options = {
+  concurrency: number;
+  deployStack: (stack: cxapi.CloudFormationStackArtifact) => Promise<void>;
+};
+
+type DeploymentState = 'pending' | 'queued' | 'deploying' | 'completed' | 'failed' | 'skipped';
+
+export const deployStacks = async (stacks: cxapi.CloudFormationStackArtifact[], { concurrency, deployStack }: Options): Promise<void> => {
+  const queue = new PQueue({ concurrency });
+  const deploymentStates = stacks.reduce((acc, stack) => ({ ...acc, [stack.id]: 'pending' as const }), {} as Record<string, DeploymentState>);
+  const deployPromises: Promise<void>[] = [];
+
+  const isStackUnblocked = (stack: cxapi.CloudFormationStackArtifact) =>
+    stack.dependencies
+      .map(({ id }) => id)
+      .filter((id) => !id.endsWith('.assets'))
+      .every((id) => deploymentStates[id] === 'completed');
+
+  const hasAnyStackFailed = (states: Record<string, DeploymentState>) => Object.values(states).includes('failed');
+
+  const enqueueStackDeploys = async () => {
+    stacks.forEach(async (stack) => {
+      if (deploymentStates[stack.id] === 'pending' && isStackUnblocked(stack)) {
+        deploymentStates[stack.id] = 'queued';
+
+        deployPromises.push(
+          queue.add(async () => {
+            // Do not start new deployments if any has already failed
+            if (hasAnyStackFailed(deploymentStates)) {
+              deploymentStates[stack.id] = 'skipped';
+              return;
+            }
+
+            deploymentStates[stack.id] = 'deploying';
+            try {
+              await deployStack(stack);
+            } catch (e) {
+              deploymentStates[stack.id] = 'failed';
+              throw e;
+            }
+
+            deploymentStates[stack.id] = 'completed';
+            await enqueueStackDeploys();
+          }),
+        );
+      }
+    });
+  };
+
+  await enqueueStackDeploys();
+
+  const results = await Promise.allSettled(deployPromises);
```

Comment:
As [rix0rrr](https://github.com/rix0rrr) mentioned `await queue.onIdle();` should probably come first here.

`Promise.allSettled` is likely to be passed only promises for stack with no dependencies (changes to the array aren't tracked after it's called) so would resolve before all stacks have completed.

Calling `await queue.onIdle();` first would resolve after everything is finished (only because deploy promises resolve after queuing stacks that become unblocked).

Also worth mentioning that `enqueueStackDeploys` is marked as async, but doesn't actually use await, same with the function passed to `stacks.forEach`

Reply:
I think that warning is because we don't immediately await the tasks returned from `queue.add`

From the p-queue docs for `.add`
> Note: If your items can potentially throw an exception, you must handle those errors from the returned Promise or they may be reported as an unhandled Promise rejection and potentially cause your process to exit immediately.

Perhaps it'd be better to have something like this:

```ts
try {
  await queue.add(...);
} catch (err) {
  deployErrors.push(err)
} 
```

Reply:
That will block/await on each individual stack deploy leading to a sequential deployment

Reply:
That code would be inside of `stacks.forEach(async stack =>)` so it wouldn't block. `Array.prototype.forEach` isn't promise aware so won't wait for completion, it'll effectively run up to the first await (starting the deployment, but not waiting for it)

Example:
![image](https://user-images.githubusercontent.com/3084997/182967041-904293c2-85a0-49c2-832f-8567a4f842a3.png)

Reply:
🤦‍♂️ thank you @ShadowDog007 I definitely was over thinking it. All appears to be working now